tommath_private.h

```diff
@@ -6,58 +6,61 @@
 
 #include "tommath.h"
 
-#ifndef MIN
-#define MIN(x, y) (((x) < (y)) ? (x) : (y))
-#endif
-
-#ifndef MAX
-#define MAX(x, y) (((x) > (y)) ? (x) : (y))
-#endif
-
 #ifdef __cplusplus
 extern "C" {
 #endif
 
+#if defined(__GNUC__) && (__GNUC__ * 100 + __GNUC_MINOR__ >= 301)
+#  define MP_DEPRECATED(x) __attribute__((deprecated("replaced by " #x)))
+#elif defined(_MSC_VER) && _MSC_VER >= 1500
+#  define MP_DEPRECATED(x) __declspec(deprecated("replaced by " #x))
+#else
+#  define MP_DEPRECATED
+#endif
+
 /* define heap macros */
-#ifndef XMALLOC
+#ifndef MP_MALLOC
```

Comment:
I like the idea, but I guess we have to keep these X-versions around until 2.0.0

Reply:
No. You just added CALLOC and removed it etc etc. This is internal API not even ABI. And these are only macros.

Reply:
Uhm, you're right... we already broke this part in #192 

The removal of `XCALLOC` doesn't break, since if someone overrides those macros with their own version it's simply not called anymore, whereas with #192 the function signatures changed and now even the name.

IMO we have to revert #192 (and part of #209) as well.

Reply:
This is not an issue. Please! This is a private header!

Reply:
I added the deprecated functions to keep the *private* ABI stable (which is already unnecessary in my opinion). Changes to tommath_private.h are a even more minor. Let's change this to move forward :)

Reply:
> Let's change this to move forward

what's so important about this change?

Reply:
why can't you just keep it for a future PR and move forward instead of discussing such stuff and then complaining that those discussions eat up your time?

btw. I will leave #192 in develop, but will revert it for 1.2.0

Reply:
We can also keep all the things as is :( However we could also try to improve things as for example in #214. This particular PR is about having a well-defined API/ABI. I consider that important.

Reply:
> We can also keep all the things as is :(

oh come on... we simply don't break existing stuff w/o warning first...

I'm fine with #214 if @karel-m is fine with it

> This particular PR is about having a well-defined API/ABI. I consider that important.

I consider that important as well.
But _renaming_ something that was there for ages, only for the sake of consistency, without any other gain, I don't consider as a such important improvement.

Reply:
> why can't you just keep it for a future PR and move forward instead of discussing such stuff and then complaining that those discussions eat up your time?

Sure I will. But then in the future I have to come back again to the PR and update it.

> btw. I will leave #192 in develop, but will revert it for 1.2.0

Ok.

Reply:
> Sure I will. But then in the future I have to come back again to the PR and update it.

if you keep it in this nice script somebody else could also do it ;-)

Reply:
You know, for that reason I made it as script. Otherwise this would be totally crazy.

Reply:
> oh come on... we simply don't break existing stuff w/o warning first...

This is not right what you are saying. I am very careful not to break stuff. I adhere to a well defined boundary of what can be changed and what cannot.

> I'm fine with #214 if @karel-m is fine with it

Ok.

> I consider that important as well. But _renaming_ something that was there for ages, only for the sake of consistency, without any other gain, I don't consider as a such important improvement.

I agree. But for example just now, mp_balance_mul was added which is actually a private function and should be called s_mp_balance_mul. By mistake, the API was extended. Such things are to be avoided.

Reply:
> This is not right what you are saying. I am very careful not to break stuff. I adhere to a well defined boundary of what can be changed and what cannot.

You're right from many aspects, but from a user perspective you're wrong. This change breaks users' implementations if they rely on the fact that they can override XMALLOC etc by defining XMALLOC at compile time of the library.

This entire stuff has been public for ages, then it has been "private" for some time, now we should make it (as possible) really private but this just needs some time (and at least a 1.2.0 release which announces the deprecation of all this stuff in 2.0.0).

> But for example just now, mp_balance_mul was added which is actually a private function and should be called s_mp_balance_mul. By mistake, the API was extended. Such things are to be avoided.

I'm talking about releases. What's in develop is not yet released. 
Yes this should've been `s_mp_balance_mul`, and we can still change it with very little effort, as it also should be `s_mp_toom_mul` etc. but we have to wait until 2.0.0 to fix this.

Which reminds me that we should start to deprecate parts of the public API as well after this PR is merged.

Reply:
@czurnieden @buggywhip what's your opinion on this?

I rethought this entire topic and am now tending towards accepting as proposed by @minad and if someone has overridden these private things, well then he has to repair it...

Reply:
regarding the deprecated API functions we're forced to keep them, as otherwise we'd have to bump the ABI

Reply:
@sjaeckel 

> am now tending towards accepting as proposed

A clean-up is necessary, no question, and we have to start at one point in time and (code)space so why not here and now?

> at least a 1.2.0 release which announces the deprecation of all this stuff in 2.0.0

Sounds good to me. And gives enough time to "round the edges".

Reply:
@sjaeckel Did you decide to merge this? If you do, ping me and I will update this PR again.

Reply:
thanks, yes please rebase

btw. would you be fine with me rebasing your commits if I sign them?

Reply:
Yes, you can also do the rebase without signing, that's ok. You are the committer.